{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4433308e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import gzip\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import os \n",
    "import matplotlib.lines as mlines\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f56ee0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SNP lists file paths\n",
    "\n",
    "craig_snp_list = pd.read_csv('/mnt/shared_folders/eResearch_glaucoma_project/Sirithi/Compare_PRS/craig_rsid_list.txt', header=None, names=['variant_id'])\n",
    "combined_snp_list = pd.read_csv('/mnt/shared_folders/eResearch_glaucoma_project/Sirithi/combining_GWAS/combined_snp_list.txt',header=None,names=['variant_id'])\n",
    "\n",
    "# Read GWAS dataset from compressed text file\n",
    "gwas_data = pd.read_csv('/mnt/shared_folders/eResearch_glaucoma_project/Sirithi/Test01_withdata/GCST90011766_buildGRCh37.txt.gz', sep='\\t', compression='gzip')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07341305",
   "metadata": {},
   "outputs": [],
   "source": [
    "POAG_clumped_snplist =  pd.read_csv('/mnt/shared_folders/eResearch_glaucoma_project/Sirithi/clumping/clumped_SNP.txt',  header=None, names=['variant_id']) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f742c834",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(combined_snp_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d025c4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find unique SNPs in each list\n",
    "unique_craig_snps = craig_snp_list[~craig_snp_list['variant_id'].isin(combined_snp_list['variant_id'])]\n",
    "unique_combined_snps = combined_snp_list[~combined_snp_list['variant_id'].isin(craig_snp_list['variant_id'])]\n",
    "\n",
    "# Find common SNPs\n",
    "common_snps = craig_snp_list[craig_snp_list['variant_id'].isin(combined_snp_list['variant_id'])]\n",
    "\n",
    "\n",
    "# Display results\n",
    "print(\"Number of unique SNPs in Craig SNP list:\", len(unique_craig_snps))\n",
    "print(\"Number of unique SNPs in Combined SNP list:\", len(unique_combined_snps))\n",
    "print(\"Number of common SNPs between Craig and Combined SNP lists:\", len(common_snps))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3859f8e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter GWAS dataset based on unique Craig SNP list\n",
    "unique_craig_snps_in_gwas = gwas_data[gwas_data['variant_id'].isin(unique_craig_snps['variant_id'])]\n",
    "\n",
    "# Display results\n",
    "print(\"Number of unique SNPs in Craig SNP list:\", len(unique_craig_snps))\n",
    "print(\"Number of unique SNPs in Craig SNP list present in GWAS dataset:\", len(unique_craig_snps_in_gwas))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edb310cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select first 10 rows\n",
    "first_10_unique_snps = unique_craig_snps_in_gwas.head(10)\n",
    "\n",
    "# Display first 10 unique SNPs\n",
    "print(\"First 10 unique SNPs in the GWAS dataset from the Craig SNP list:\")\n",
    "print(first_10_unique_snps)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6689c88c",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(gwas_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4aaa106c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter GWAS dataset based on SNP lists\n",
    "snp_positions_craig = gwas_data[gwas_data['variant_id'].isin(craig_snp_list['variant_id'])][['chromosome', 'base_pair_location']]\n",
    "snp_positions_combined = gwas_data[gwas_data['variant_id'].isin(combined_snp_list['variant_id'])][['chromosome', 'base_pair_location']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1710fd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Comparing SNP p-values\n",
    "\n",
    "# Drop rows with null values in the 'chromosome' column\n",
    "gwas_data = gwas_data.dropna(subset=['chromosome'])\n",
    "\n",
    "# Remove chromosome 23 if it has no data\n",
    "chromosome_counts = gwas_data['chromosome'].value_counts()\n",
    "chromosomes_with_data = chromosome_counts[chromosome_counts > 0].index\n",
    "gwas_data = gwas_data[gwas_data['chromosome'].isin(chromosomes_with_data)]\n",
    "\n",
    "# Convert chromosome to categorical and sort by chromosome\n",
    "gwas_data['chromosome'] = gwas_data['chromosome'].astype('category')\n",
    "gwas_data['chromosome'] = gwas_data['chromosome'].cat.set_categories(gwas_data['chromosome'].unique(), ordered=True)\n",
    "gwas_data = gwas_data.sort_values('chromosome')\n",
    "\n",
    "# Calculate -log10(p-value)\n",
    "gwas_data['minuslog10pvalue'] = -np.log10(gwas_data['p_value'])\n",
    "\n",
    "# Add a new column to differentiate MTAG and POAG SNPs\n",
    "gwas_data['category'] = 'Other'\n",
    "gwas_data.loc[gwas_data['variant_id'].isin(craig_snp_list['variant_id']), 'category'] = 'MTAG SNPs'\n",
    "gwas_data.loc[gwas_data['variant_id'].isin(combined_snp_list['variant_id']), 'category'] = 'Full-PRS SNPs'\n",
    "\n",
    "# Filter data to only include MTAG and POAG SNPs\n",
    "filtered_gwas_data = gwas_data[gwas_data['category'].isin(['MTAG SNPs', 'Full-PRS SNPs'])]\n",
    "\n",
    "# Custom color palette\n",
    "custom_palette = {\n",
    "    'MTAG SNPs': '#87CEEB',  # Sky Blue\n",
    "    'Full-PRS SNPs': 'darkorange'   # Coral\n",
    "}#87CEEB\n",
    "\n",
    "# Create a violin plot split by chromosome and category\n",
    "plt.figure(figsize=(12, 8))\n",
    "sns.violinplot(x='chromosome', y='minuslog10pvalue', hue='category', data=filtered_gwas_data, split=True, cut=0, inner=None, palette=custom_palette, alpha = 0.9)\n",
    "\n",
    "# Add a dotted line at -log10(0.001) which is 3\n",
    "plt.axhline(y=3, color='gray', linestyle='--')\n",
    "\n",
    "# Add text label for the dotted line\n",
    "plt.text(len(gwas_data['chromosome'].unique()) - 1.5, 3, 'p-value 0.001', color='black', ha='right', va='bottom',fontsize = 12)\n",
    "\n",
    "# Set axis limits\n",
    "plt.ylim([0, 4])  # Adjust the y-axis limit as needed\n",
    "\n",
    "# x-axis label\n",
    "plt.xlabel('Chromosome',fontsize = 12)\n",
    "\n",
    "# y-axis label\n",
    "plt.ylabel('-log10(p-value)',fontsize = 12 )\n",
    "\n",
    "# Modify legend title\n",
    "plt.legend(title='Category')\n",
    "\n",
    "# Display the combined figure\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save the figure\n",
    "folder_path = '/mnt/shared_folders/eResearch_glaucoma_project/Sirithi/Graph_Images/facet_plots/'\n",
    "file_path = os.path.join(folder_path, 'pval_comparison_mtagpoag.pdf')\n",
    "\n",
    "# Ensure the folder exists and save the figure\n",
    "os.makedirs(folder_path, exist_ok=True)\n",
    "plt.savefig(file_path)\n",
    "\n",
    "plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "446f532c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# Add a new column to differentiate Craig and POAG SNPs in gwas_data\n",
    "gwas_data['source'] = 'Other'\n",
    "gwas_data.loc[gwas_data['variant_id'].isin(craig_snp_list['variant_id']), 'source'] = 'Craig SNPs'\n",
    "gwas_data.loc[gwas_data['variant_id'].isin(POAG_clumped_snplist['variant_id']), 'source'] = 'POAG SNPs'\n",
    "\n",
    "# Filter data to only include Craig and POAG SNPs\n",
    "filtered_gwas_data = gwas_data[gwas_data['source'].isin(['Craig SNPs', 'POAG SNPs'])].copy()\n",
    "\n",
    "# Convert chromosome to categorical and sort by chromosome\n",
    "filtered_gwas_data.loc[:, 'chromosome'] = filtered_gwas_data['chromosome'].astype('category')\n",
    "filtered_gwas_data.loc[:, 'chromosome'] = filtered_gwas_data['chromosome'].cat.set_categories(filtered_gwas_data['chromosome'].unique(), ordered=True)\n",
    "filtered_gwas_data = filtered_gwas_data.sort_values('chromosome')\n",
    "\n",
    "# Create a violin plot split by chromosome and source\n",
    "plt.figure(figsize=(14, 8))\n",
    "sns.violinplot(x='chromosome', y='base_pair_location', hue='source', data=filtered_gwas_data, split=True, cut=0, inner=None)\n",
    "\n",
    "# x-axis label\n",
    "plt.xlabel('Chromosome',fontsize= 16)\n",
    "\n",
    "# y-axis label\n",
    "plt.ylabel('Base Pair Location', fontsize= 16)\n",
    "\n",
    "#\n",
    "\n",
    "# Show the graph\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5077a73c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count the total number of SNPs shown in the dot plot\n",
    "total_snps = len(filtered_gwas_data)\n",
    "print(f\"Total number of SNPs shown in the dot plot: {total_snps}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b9f9374",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Comparing base-pair locations\n",
    "\n",
    "# Add a new column to differentiate Craig and POAG SNPs in gwas_data\n",
    "gwas_data['source'] = 'Other'\n",
    "gwas_data.loc[gwas_data['variant_id'].isin(craig_snp_list['variant_id']), 'source'] = 'Craig SNPs'\n",
    "gwas_data.loc[gwas_data['variant_id'].isin(POAG_clumped_snplist['variant_id']), 'source'] = 'Full-PRS SNPs'\n",
    "\n",
    "# Filter data to only include Craig and POAG SNPs\n",
    "filtered_gwas_data = gwas_data[gwas_data['source'].isin(['Craig SNPs', 'Full-PRS SNPs'])]\n",
    "\n",
    "# Convert chromosome to categorical and sort by chromosome\n",
    "filtered_gwas_data['chromosome'] = filtered_gwas_data['chromosome'].astype('category')\n",
    "filtered_gwas_data['chromosome'] = filtered_gwas_data['chromosome'].cat.set_categories(\n",
    "    filtered_gwas_data['chromosome'].unique(), ordered=True)\n",
    "filtered_gwas_data = filtered_gwas_data.sort_values('chromosome')\n",
    "\n",
    "custom_palette = {'Craig SNPs': '#87CEEB',   # Light Blue\n",
    "                  'Full-PRS SNPs': '#FFA500'}  \n",
    "\n",
    "# Create the plot\n",
    "plt.figure(figsize=(12, 8))\n",
    "ax = sns.stripplot(x='chromosome', y='base_pair_location', hue='source', data=filtered_gwas_data,\n",
    "                   dodge=True, jitter=True, palette=custom_palette, marker='o', size=3, alpha=0.7)\n",
    "\n",
    "# Custom legend markers with larger size\n",
    "mtag_legend = mlines.Line2D([], [], color='#87CEEB', marker='o', linestyle='None', markersize=5, label='MTAG SNPs')\n",
    "full_prs_legend = mlines.Line2D([], [], color='#FFA500', marker='o', linestyle='None', markersize=5, label='Full-PRS SNPs')\n",
    "\n",
    "# Add the custom legend\n",
    "ax.legend(handles=[mtag_legend, full_prs_legend], title='')\n",
    "\n",
    "# x-axis label\n",
    "plt.xlabel('Chromosome', fontsize=12)\n",
    "\n",
    "# y-axis label\n",
    "plt.ylabel('Base Pair Location', fontsize=12)\n",
    "\n",
    "# Display the combined figure\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save the figure\n",
    "folder_path = '/mnt/shared_folders/eResearch_glaucoma_project/Sirithi/Graph_Images/facet_plots/'\n",
    "file_path = os.path.join(folder_path, 'BP_location_comparison_mtagpoag.pdf')\n",
    "\n",
    "# Ensure the folder exists and save the figure\n",
    "os.makedirs(folder_path, exist_ok=True)\n",
    "plt.savefig(file_path)\n",
    "\n",
    "plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1e51aea",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
